### ISSUE TITLE
Simple query execution blocked in TPCH dataset 

### ISSUE THREAD

**Issue by markjin1990** [facebookincubator/velox#1634]
**Problem Description**
This is a follow-up of the issue we discussed on 05/12 weekly meeting when presenting our TPCH benchmark results. The query execution of a simple SQL,
`SELECT Min(ps.supplycost) FROM tpch.partsupp AS ps, tpch.supplier AS s WHERE s.suppkey = ps.suppkey AND s.name = 'Supplier#002301894'`, on the tpch dataset got blocked in the "partsupp" table scan stage (Stage 2 in the screenshot below) for > 1 hour and could not finish in our internal PrestoCpp setup.
![image](https://user-images.githubusercontent.com/1387708/168784840-32ed55be-cefd-4d3d-848b-4d39432bfeec.png)

Most workers have the same log as follows, suggesting the worker periodically checks a certain output buffer but could not find anything
```sh
I0517 10:36:14.985965   180 TaskManager.cpp:463] TaskManager::getResults 20220517_103218_00026_sebte.2.0.13, 11, 9
I0517 10:36:16.988066   180 TaskManager.cpp:463] TaskManager::getResults 20220517_103218_00026_sebte.2.0.13, 11, 9
I0517 10:36:18.990137   180 TaskManager.cpp:463] TaskManager::getResults 20220517_103218_00026_sebte.2.0.13, 11, 9
I0517 10:36:20.991004   180 TaskManager.cpp:463] TaskManager::getResults 20220517_103218_00026_sebte.2.0.13, 11, 9
I0517 10:36:22.994086   180 TaskManager.cpp:463] TaskManager::getResults 20220517_103218_00026_sebte.2.0.13, 11, 9
```

Further, we found out that the table scan stage (Stage 2) got stuck because the output buffer size is greater than  'continueSize_' (defined at
https://github.com/facebookincubator/velox/blob/main/velox/exec/PartitionedOutputBufferManager.h#L172) and all producers therefore got blocked.

A quick way to verify our discovery is that if we increase `maxSize_` to a larger value (say 1G), the query execution could quickly finish in minutes (by default, continueSize_ is set to be 90% of maxSize_ and maxSize_ is set to 32MB). 

We will continue to look into this issue but we would very much appreciate if anyone could shed some light on what might cause the output buffer size so full to block the producers.

**Some Background info**
We are testing on a 1TB TPCH dataset stored in remote hdfs. 

Worker Size: (20Core, 96GB) x 30
Worker Configs
- Use_mmap_allocator
- system_memory_gb=80g
- num_query_threads=64
- max_block_value_set_length=512
- velox_memory_pool_mb=10240
- file_handle_cache_mb=2048
- num_io_threads=64
- task.max-drivers-per-task=40

Logical plan

```sh
 - Output[_col0] => [min:double]                                                                                                                                                                                
         _col0 := min (1:16)                                                                                                                                                                                    
     - Aggregate(FINAL) => [min:double]                                                                                                                                                                         
             min := "presto.default.min"((min_9)) (1:16)                                                                                                                                                        
         - LocalExchange[SINGLE] () => [min_9:double]                                                                                                                                                           
             - RemoteStreamingExchange[GATHER] => [min_9:double]                                                                                                                                                
                 - Aggregate(PARTIAL) => [min_9:double]                                                                                                                                                         
                         min_9 := "presto.default.min"((supplycost)) (1:16)                                                                                                                                     
                     - InnerJoin[("suppkey" = "suppkey_0")] => [supplycost:double]                                                                                                                              
                             Distribution: PARTITIONED                                                                                                                                                          
                         - RemoteStreamingExchange[REPARTITION] => [suppkey:bigint, supplycost:double]                                                                                                          
                                 Estimates: {rows: ? (?), cpu: ?, memory: 0.00, network: ?}                                                                                                                     
                             - TableScan[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=partsupp, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.partsupp{}]'}] => [suppkey:bigint, supplycost:double]
                                     Estimates: {rows: ? (?), cpu: ?, memory: 0.00, network: 0.00}                                                                                                              
                                     LAYOUT: tpch.partsupp{}                                                                                                                                 
                                     supplycost := supplycost:double:3:REGULAR (1:40)                                                                                                                           
                                     suppkey := suppkey:bigint:1:REGULAR (1:40)                                                                                                                                 
                         - LocalExchange[HASH] (suppkey_0) => [suppkey_0:bigint]                                                                                                                                
                                 Estimates: {rows: ? (?), cpu: ?, memory: 0.00, network: ?}                                                                                                                     
                             - RemoteStreamingExchange[REPARTITION] => [suppkey_0:bigint]                                                                                                                       
                                     Estimates: {rows: ? (?), cpu: ?, memory: 0.00, network: ?}                                                                                                                 
                                 - TableScan[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=tpch, tableName=supplier, analyzePartitionValues=Optional.empty}', layout='Optional[tpch.supplier{domains={name=[ [["Supplier#002301894"]] ]}}]'}] => [suppkey_0:bigint]
                                         Estimates: {rows: ? (?), cpu: ?, memory: 0.00, network: 0.00}                                                                                                          
                                         LAYOUT: tpch.supplier{domains={name=[ [["Supplier#002301894"]] ]}}                                                                                  
                                         suppkey_0 := suppkey:bigint:0:REGULAR (1:85)                                                                                                                           
                                         name:string:1:REGULAR                                                                                                                                                  
                                             :: [["Supplier#002301894"]]              
```

**Update [06/21]: We believe that we have found the root cause for this issue.** 

### Root Cause
In a nutshell, the SQL execution is blocked when some of the downstream tasks consuming the data from the outputBuffers of the upstream task complete & abort early without notifying the upstream task. All producers in the upstream task will keep dumping data into outputBuffers until they are blocked by `maxSize_` and `continueSize_` and may not be unblocked because most buffers cannot be consumed and freed up.

It is difficult to reproduce this problem in a local setup, but it is easy to demonstrate the root cause using the following simplified example about Stage 1 and Stage 2 from the sample SQL we provide in the issue description section.

![performance issue illustration (1)](https://user-images.githubusercontent.com/1387708/174884186-393eaf5a-550a-40b8-bfa1-141cff55b4ef.png)

1. The downstream tasks in Stage 1 are reading from the output result of upstream tasks in Stage 2. 
2. If task 1.0.2 and 1.0.3 complete early (their hashProbe learn that no result in their assigned partitions from Stage 2 needs to be joined after the result of Stage 3 returns), they will stop consuming data from all tasks in Stage 2.
3. However, none of the tasks in Stage 2 knows that Task 1.0.2 and Task 1.0.3 have completed and all producers in Stage 2 will keep reading data and dumping to corresponding OutputBuffers. 
4. Note that constants `maxSize_` and `continueSize_` limit the overall size of all buffers for a given task on a worker node. The `totalSize_` of PartitionedOutputBuffer will quickly exceeds `maxSize_`, and hence all producers all blocked.
5. Now, only Task 1.0.1 is alive in Stage 1, and it only reads from the buffer in the destination 1 in PartitionOutputBuffer of all tasks in Stage 2. Although, the buffer at destination 1 can be consumed and freed up, when all other buffers (at destination 2 and 3) are not consumed or freed up, 
  - the `totalSize_` may still be bigger than `continueSize_`,  all producers are still blocked, 
  - or the `totalSize_` might become lower than `continueSize_` after some pages at definition 1 are freed up, all producers are unblocked and immediately start a race reading and dumping data into OutputBuffers. If producers for destination 2 and destination 3 quickly fill up PartitionedOutputBuffer, then the producer for destination 1 will be blocked
6. In either way, the task 1.0.1 and the overall SQL is almost guaranteed to be blocked eventually.

### Proposed Solution
We have implemented a fix in our internal repo and successfully unblocked the sample SQL in the current issue as well as all other tpch SQLs that were blocking before.  Our high-level idea is simply sending an `abortResults` request to all data sources (upstream tasks) when a task completes. The actual fix is not as trivial as it seems and we can create a PR later to showcase.

**Comment by mbasmanova** [comment]
@markjin1990 Have you tried reproducing this problem in a e2e test?

**Comment by markjin1990** [comment]
> @markjin1990 Have you tried reproducing this problem in a e2e test?

Will give it a try and post updates :)

**Comment by markjin1990** [comment]
**Update: We believe that we have found the root cause for this issue.** 
### Root Cause
In a nutshell, the SQL execution is blocked when some of the downstream tasks consuming the data from the outputBuffers of the upstream task complete & abort early without notifying the upstream task and all producers in the upstream task will keep dumping data into outputBuffers until they are blocked by `maxSize_` and `continueSize_` and may not be unblocked because most buffers cannot be consumed and freed up.

It is difficult to reproduce this problem in a local setup, but it is easy to demonstrate the root cause using the following simplified example about Stage 1 and Stage 2 from the sample SQL we provide in the issue description section.

![performance issue illustration (1)](https://user-images.githubusercontent.com/1387708/174884186-393eaf5a-550a-40b8-bfa1-141cff55b4ef.png)

1. The downstream tasks in Stage 1 are reading from the output result of upstream tasks in Stage 2. 
2. If task 1.0.2 and 1.0.3 complete early (their hashProbe learn that no result in their assigned partitions from Stage 2 needs to be joined after the result of Stage 3 returns), they will stop consuming data from all tasks in Stage 2.
3. However, none of the tasks in Stage 2 knows that Task 1.0.2 and Task 1.0.3 have completed and all producers in Stage 2 will keep reading data and dumping to corresponding OutputBuffers. 
4. Note that constants `maxSize_` and `continueSize_` limit the overall size of all buffers for a given task on a worker node. The `totalSize_` of PartitionedOutputBuffer will quickly exceeds `maxSize_`, and hence all producers all blocked.
5. Now, only Task 1.0.1 is alive in Stage 1, and it only reads from the buffer in the destination 1 in PartitionOutputBuffer of all tasks in Stage 2. Although, the buffer at destination 1 can be consumed and freed up, when all other buffers (at destination 2 and 3) are not consumed or freed up, 
  - the `totalSize_` may still be bigger than `continueSize_`,  all producers are still blocked, 
  - or the `totalSize_` might become lower than `continueSize_` after some pages at definition 1 are freed up, all producers are unblocked and immediately start a race reading and dumping data into OutputBuffers. If producers for destination 2 and destination 3 quickly fill up PartitionedOutputBuffer, then the producer for destination 1 will be blocked
6. In either way, the task 1.0.1 and the overall SQL is almost guaranteed to be blocked eventually.

### Proposed Solution
We have implemented a fix in our internal repo and successfully unblocked the sample SQL in the current issue as well as all other tpch SQLs that were blocking before.  Our high-level idea is simply sending an `abortResults` request to all data sources (upstream tasks) when a task completes. The actual fix is not as trivial as it seems and we can create a PR later to showcase.

**Comment by mbasmanova** [comment]
> Update [06/21]: We believe that we have found the root cause for this issue.

@markjin1990 Mark, thank you for the update. Looking forward to the PR with the fix.

**Comment by mbasmanova** [comment]
I wonder if this is a similar: https://github.com/facebookincubator/velox/pull/1332

**Comment by frankobe** [comment]
@mbasmanova The root cause is similar that the exchange source is not closed to free the memory as the downstream stage is terminated early. However your fix is for local exchange, we will fix it for remote exchange 

> I wonder if this is a similar: #1332

**Comment by frankobe** [comment]
@tanjialiang You can try the following two PRs to fix the remote exchange
https://github.com/facebookincubator/velox/pull/1886
https://github.com/facebookexternal/presto_cpp/pull/789

**Comment by mbasmanova** [comment]
Another query that hits the same problem:

```
select count(*) 
from customer_3k c, nation_3k n, region_3k r 
where c.nationkey =  n.nationkey and r.regionkey = n.regionkey and r.name = 'ASIA';
```

**Comment by mbasmanova** [comment]
Some prep work to enable fixing this problem: #2379

**Comment by mbasmanova** [comment]
@frankobe Frank, this problem should be fixed now. The fix has 3 parts: #2379, #2380, #2384 and applies to all types of exchanges including broadcast, partitioned and merge (exchange of sorted data). I'm going to close #1886.
